cloudevents/verifiability-spec.md

```diff
@@ -0,0 +1,210 @@
+        * Producer and consumer must agree on an appropriate type
+
+## Verifiability Implementations
+
+While this proposal is agnostic to the specific implementations, we set the following rules for verifiability implementations:
+
+A verifiability implementation **MUST**:
+
+* transport the verification material in the same message as the event
+* use the two relevant context attributes in this proposal
+
+A verifiability implementation **SHOULD**:
+
+* avoid depending on canonicalization (and its [unnecessarily large attack surface](https://github.com/secure-systems-lab/dsse/blob/master/background.md#motivation))
+* transport the verification material detached from the event payload (this helps avoid canonicalization)
+
+A verifiability implementation **MUST FAIL** on:
+
+* Duplication of a relevant context attribute (e.g. two `verificationmaterial` attributes being set in a message)
+* One relevant context attribute being present but the other missing (e.g. `verificationmaterial` being set but `verificationmaterialtype` missing)
+* Unknown material type (e.g. the implementation may only attempt to use the verification material if it knows how to do so)
+* Invalid material (e.g. the verification material did not match the received event)
+
+In order to get a verifiability implementation into the official CloudEvents SDK, the following requirements are set.
+
+## Verifiability Implementation Proposals
+
+A proposal for adding a verifiability implementation to the CE SDK **MUST:**
+
+* provide specs on how to implement verification
+* provide test vectors
+* provide a verification coverage table (see Verification coverage table)
+
+Test vectors ensure that implementations are consistent which does NOT affect verifiability. 
+
+## Example
+
+To illustrate, let’s walk through how to implement a verification implementation CloudEvents provide test vectors as well as a verification coverage table.
+
+### Verification Implementation
+
+This is an example message containing a CloudEvent directly from the spec. It happens to be an HTTP structured mode message in JSON format. Our goal is to come up with the most insecure but also most concise verification implementation imaginable: one based on an md5 digest!
+
+Here is how a message containing that CloudEvent might look:
+
+```
+content-length: 209
+content-type: application/json
+
+{
+    "specversion" : "1.0",
+    "type" : "com.example.someevent",
+    "source" : "/mycontext",
+    "subject": null,
+    "id" : "D234-1234-1234",
+    "time" : "2018-04-05T17:31:00Z",
+    "data" : "I'm just a string"
+}
+```
+
+In order to make this event verifiable, an md5 digest of the body of the message is computed: `dc98e72d0f9980e89af5f8e7692d58d7`. This value will then be base64 encoded and added as the `verificationmaterial` context attribute:
+
+```
+content-length: 209
+content-type: application/json
+ce-verificationmaterial: ZGM5OGU3MmQwZjk5ODBlODlhZjVmOGU3NjkyZDU4ZDc=
```

Comment:
these should appear in the body along-side the other CE attributes, no?

Reply:
Depends on if it is structured or binary mode.

By along-side, you are meaning within the payload?

Reply:
Oh I see what you mean. Yes, that's a mistake on our end. I'll update after today's SIG.